R/guides-.r

```diff
@@ -76,303 +76,473 @@ guides <- function(...) {
+
+    guides <- gtable_add_grob(
+      guides, background,
+      t = 1, l = 1, b = -1, r = -1,
+      z = -Inf, clip = "off",
+      name = "legend.box.background"
+    )
+    guides$name <- "guide-box"
+    guides
+  },
+
+  # building non-position guides - called in ggplotGrob (plot-build.r)
+  #
+  # the procedure is as follows:
+  #
+  # 1. guides$setup()
+  #      generates a guide object for every scale-aesthetic pair
+  #
+  # 2. guides$train()
+  #      train each scale and generate guide definition for all guides
+  #      here, one guide object for one scale
+  #
+  # 2. guides$merge()
+  #      merge guide objects if they are overlayed
+  #      number of guide objects may be less than number of scales
+  #
+  # 3. guides$process_layers()
+  #      process layer information and generate geom info.
+  #
+  # 4. guides$draw()
+  #      generate guide grob from each guide object
+  #      one guide grob for one guide object
+  #
+  # 5. guides$assemble()
+  #      arrange all guide grobs
+
+  build = function(self, scales, layers, default_mapping,
+                   position, theme, labels) {
+
+    position  <- legend_position(position)
+    no_guides <- zeroGrob()
+    if (position == "none") {
+      return(no_guides)
     }
 
-    guides <- gtable_col(name = "guides",
-      grobs = ggrobs,
-      width = max(widths), heights = heights)
-
-    # add space between the guide-boxes
-    guides <- gtable_add_row_space(guides, theme$legend.spacing.y)
-  }
-
-  # Add margins around the guide-boxes.
-  theme$legend.box.margin <- theme$legend.box.margin %||% margin()
-  guides <- gtable_add_cols(guides, theme$legend.box.margin[4], pos = 0)
-  guides <- gtable_add_cols(guides, theme$legend.box.margin[2], pos = ncol(guides))
-  guides <- gtable_add_rows(guides, theme$legend.box.margin[1], pos = 0)
-  guides <- gtable_add_rows(guides, theme$legend.box.margin[3], pos = nrow(guides))
+    theme$legend.key.width  <- theme$legend.key.width  %||% theme$legend.key.size
+    theme$legend.key.height <- theme$legend.key.height %||% theme$legend.key.size
 
-  # Add legend box background
-  background <- element_grob(theme$legend.box.background %||% element_blank())
 
-  guides <- gtable_add_grob(guides, background, t = 1, l = 1,
-    b = -1, r = -1, z = -Inf, clip = "off", name = "legend.box.background")
-  guides$name <- "guide-box"
-  guides
-}
-
-# Generics ----------------------------------------------------------------
-
-#' S3 generics for guides.
-#'
-#' You will need to provide methods for these S3 generics if you want to
-#' create your own guide object. They are currently undocumented; use at
-#' your own risk!
-#'
-#' @param guide The guide object
-#' @keywords internal
-#' @name guide-exts
-NULL
-
-#' @export
-#' @rdname guide-exts
-guide_train <- function(guide, scale, aesthetic = NULL) UseMethod("guide_train")
+    direction <- if (position == "inside") "vertical" else position
```

Comment:
Does this mean that if the legend is placed inside the panel it will always be vertical? Is there any sort of reasoning for this or are you just porting old behaviour?

Reply:
no - a rename to `default_direction` would be fine with me